src/chain.h

```diff
@@ -300,13 +308,23 @@ class CBlockIndex
         return ((nStatus & BLOCK_VALID_MASK) >= nUpTo);
     }
 
+    //! @returns true if the block is assumed-valid; probably means it is queued to be
+    //!   validated by a background chainstate.
+    bool IsAssumedValid() const { return nStatus & BLOCK_ASSUMED_VALID; }
+
     //! Raise the validity level of this block index entry.
     //! Returns true if the validity was changed.
     bool RaiseValidity(enum BlockStatus nUpTo)
     {
         assert(!(nUpTo & ~BLOCK_VALID_MASK)); // Only validity flags allowed.
-        if (nStatus & BLOCK_FAILED_MASK)
-            return false;
+        if (nStatus & BLOCK_FAILED_MASK) return false;
+
+        // If this block had been marked assumed-valid and we're raising
+        // its validity to a certain point, there is no longer an assumption.
+        if (nStatus & BLOCK_ASSUMED_VALID && nUpTo >= BLOCK_VALID_SCRIPTS) {
```

Comment:
212b80dad perhaps use the bool helper added in this commit?
```suggestion
        if (IsAssumedValid() && nUpTo >= BLOCK_VALID_SCRIPTS) {
```

Reply:
Given I'm unsetting the same flag just below I think this could be somewhat confusing.